src/vertex_tools.rs

```diff
@@ -0,0 +1,398 @@
+#[allow(dead_code)]
+pub fn ray_intersects_triangle(
+    ray_origin: &Point3<f32>,
+    ray_vector: &Vector3<f32>,
+    triangle_vertices: (&Point3<f32>, &Point3<f32>, &Point3<f32>),
+) -> Option<Point3<f32>> {
+    let edge_1_vector = triangle_vertices.1 - triangle_vertices.0;
+    let edge_2_vector = triangle_vertices.2 - triangle_vertices.0;
+    let perpendicular_vector = ray_vector.cross(&edge_2_vector);
+    let determinant = edge_1_vector.dot(&perpendicular_vector);
+    // This ray is parallel to this triangle.
+    if approx::relative_eq!(determinant, 0.0) {
+        return None;
+    }
+    let inverse_determinant = 1.0 / determinant;
+    let tangent_vector = ray_origin - triangle_vertices.0;
+    let u_parameter = inverse_determinant * tangent_vector.dot(&perpendicular_vector);
+    if u_parameter < 0.0 || u_parameter > 1.0 {
+        return None;
+    }
+    let q_vector = tangent_vector.cross(&edge_1_vector);
+    let v_parameter = inverse_determinant * ray_vector.dot(&q_vector);
+    if v_parameter < 0.0 || u_parameter + v_parameter > 1.0 {
+        return None;
+    }
+    // At this stage we can compute t_parameter to find out where the
+    // intersection point is on the line.
+    let t_parameter = inverse_determinant * edge_2_vector.dot(&q_vector);
+    // Ray-triangle intersection
+    if t_parameter > f32::EPSILON && t_parameter < 1.0 / f32::EPSILON {
+        Some(ray_origin + ray_vector * t_parameter)
+    } else {
+        None
+    }
+}
+
+#[allow(dead_code)]
+pub struct PointOnLine {
+    clamped: Point3<f32>,
+    unclamped: Point3<f32>,
+}
+
+/// Pulls arbitrary point to an arbitrary line.
+///
+/// Returns both, a clamped point on the line adn an unclamped points on an
+/// endless ray.
+///
+/// https://stackoverflow.com/questions/3120357/get-closest-point-to-a-line
+pub fn pull_point_to_line(
+    point: &Point3<f32>,
+    line_points: (&Point3<f32>, &Point3<f32>),
+) -> PointOnLine {
+    let line_start = line_points.0;
+    let line_end = line_points.1;
+    let start_to_point = point - line_start;
+    let start_to_end = line_end - line_start;
+
+    let square_magnitude_start_to_end = nalgebra::distance_squared(line_start, line_end);
+    let start_to_point_dot_start_to_end = start_to_point.dot(&start_to_end);
+    let line_parameter = start_to_point_dot_start_to_end / square_magnitude_start_to_end;
+
+    let closest_point_on_line_unclamped =
+        Point3::from(line_start.coords.lerp(&line_end.coords, line_parameter));
+    let closest_point_on_line_clamped = if line_parameter < 0.0 {
+        *line_start
+    } else if line_parameter > 1.0 {
+        *line_end
+    } else {
+        closest_point_on_line_unclamped
+    };
+
+    PointOnLine {
+        clamped: closest_point_on_line_clamped,
+        unclamped: closest_point_on_line_unclamped,
+    }
+}
+
+#[allow(dead_code)]
+pub struct PulledPoint {
```

Comment:
I disagree with the naming. This can be any pulled point, also there is a convention for a  *point on mesh*, which contains absolute coordinates and geometry-relative coordinates (in our case they would be the face index and barycentric coords).